src/shared/logger.ts

```diff
@@ -0,0 +1,221 @@
+        error: (...message: ErrorOrString[]) =>
+            writeToLogs(generateWriteParams(newLogger, 'error', message, params.outputChannel)),
+        releaseLogger: () => releaseLogger(newLogger)
+    }
+}
+
+function getDefaultLogPath(): string {
+    if (os.platform() === 'win32') {
+        return path.join(os.homedir(), 'AppData', 'Roaming', LOG_RELATIVE_PATH, DEFAULT_LOG_NAME)
+    } else if (os.platform() === 'darwin') {
+        return path.join(os.homedir(), 'Library', 'Application Support', LOG_RELATIVE_PATH, DEFAULT_LOG_NAME)
+    } else {
+        return path.join(os.homedir(), '.config', LOG_RELATIVE_PATH, DEFAULT_LOG_NAME)
+    }
+}
+
+async function openLogFile(): Promise<void> {
+    if (defaultLogger.logPath) {
+        await vscode.window.showTextDocument(vscode.Uri.file(path.normalize(defaultLogger.logPath)))
+    }
+}
+
+function releaseLogger(logger: winston.Logger): void {
+    logger.clear()
+}
+
+function formatMessage(level: string, message: ErrorOrString[]): string {
+    let final: string = `${makeDateString('logfile')} [${level.toUpperCase()}]:`
+    for (const chunk of message) {
+        if (chunk instanceof Error) {
+            final = `${final} ${chunk.stack}`
+        } else {
+            final = `${final} ${chunk}`
+        }
+    }
+
+    return final
+}
+
+function writeToLogs(params: WriteToLogParams): void {
+    const message = formatMessage(params.level, params.message)
+    params.logger.log(params.level, message)
+    if (params.outputChannel) {
+        writeToOutputChannel(params.logger.levels[params.level],
+                             params.logger.levels[params.logger.level],
+                             message,
+                             params.outputChannel)
+    }
+}
+
+function writeToOutputChannel(messageLevel: number,
+                              logLevel: number,
+                              message: string,
+                              outputChannel: vscode.OutputChannel): void {
+    // using default Winston log levels (mapped to numbers): https://github.com/winstonjs/winston#logging
+    if (messageLevel <= logLevel) {
+        outputChannel.appendLine(message)
+    }
+}
+
+// matches VS Code's log file name format
+// YYYYMMDDThhmmss (note the 'T' prior to time)
+// Uses local timezone
+function makeDateString(type: 'filename' | 'logfile'): string {
+    const d = new Date()
+    const isFilename: boolean = type === 'filename'
+
+    return `${d.getFullYear()}${chooseSpacer(isFilename, '', '-')}` +
+    // String.prototype.padStart() is not available in Typescript...
+    `${padNumber(d.getMonth() + 1)}${chooseSpacer(isFilename, '', '-')}` +
+    `${padNumber(d.getDate())}${chooseSpacer(isFilename, 'T', ' ')}` +
+    `${padNumber(d.getHours())}${chooseSpacer(isFilename, '', ':')}` +
+    `${padNumber(d.getMinutes())}${chooseSpacer(isFilename, '', ':')}` +
+    `${padNumber(d.getSeconds())}${chooseSpacer(isFilename, '', ':')}`
```

Comment:
Remove this last `chooseSpacer` call to remove the trailing colon:
`2019-03-01 14:53:12: [ERROR]: InvalidClientTokenId: The security token included in the request is invalid.`